packages/seacas/libraries/ioss/src/Ioss_DynamicTopology.C

```diff
@@ -36,14 +36,463 @@
+    Ioss::FileInfo file = Ioss::FileInfo(full_filename);
+    exists = file.exists();
+  }
+
+  if (is_parallel) {
+    int iexists = exists ? 1 : 0;
+    util.broadcast(iexists, 0);
+    exists = iexists == 1;
+  }
+  return exists;
+}
+
+int file_exists( const Ioss::ParallelUtils &util,
+                 const std::string& filename,
+                 std::string& message,
+                 bool specifiedDecomp )
+{
+  std::string filenameBase = filename;
+  const int par_size = util.parallel_size();
+  const int par_rank = util.parallel_rank();
+
+  if( par_size > 1 && !specifiedDecomp ) {
+    filenameBase = Ioss::Utils::decode_filename(filenameBase, par_rank, par_size);
+  }
+
+  Ioss::FileInfo file = Ioss::FileInfo(filenameBase);
+  return file.parallel_exists(util.communicator(), message);
+}
+
+bool internal_decomp_specified(const Ioss::PropertyManager& props)
+{
+  bool internalDecompSpecified = false;
+
+  const std::string restartDecompMethod("RESTART_DECOMPOSITION_METHOD");
+
+  if (props.exists(restartDecompMethod)) {
+    Ioss::Property prop = props.get(restartDecompMethod);
+    if (prop.get_string() != "EXTERNAL") {
+      internalDecompSpecified = true;
+    }
+  }
+
+  return internalDecompSpecified;
+}
+
+using FileNameGenerator = std::function<std::string(const std::string& baseFileName, unsigned step)>;
+
+std::pair<std::string, Ioss::DatabaseIO*>
+expand_topology_files(FileNameGenerator generator,
+                      const Ioss::ParallelUtils &util,
+                      const std::string& basename, const std::string& db_type,
+                      const Ioss::PropertyManager& properties, int step)
+{
+  // See if there are multiple topology files
+
+  // If the file exists on all processors, return the filename.
+  // If the file does not exist on any processors, return "";
+  // If the file exists on some, but not all, throw an exception.
+
+  std::string filename = generator(basename, step);
+  Ioss::DatabaseIO* db = nullptr;
+
+  bool internalDecompSpecified = internal_decomp_specified(properties);
+  std::string message;
+  int exists = ::file_exists(util, filename, message, internalDecompSpecified);
+
+  int par_size = util.parallel_size();
+  int par_rank = util.parallel_rank();
+
+  if( exists > 0 && exists < par_size ) {
+    std::ostringstream errmsg;
+    errmsg << "ERROR: Unable to open input database";
+    if(par_rank == 0) {
+      errmsg << " '" << filename << "'" << "\n\ton processor(s): " << message;
+    } else {
+      errmsg << ". See processor 0 output for more details.\n";
+    }
+    IOSS_ERROR(errmsg);
+  }
+
+  if( exists == par_size ) {
+    db = Ioss::IOFactory::create(db_type, filename, Ioss::QUERY_TIMESTEPS_ONLY,
+                                 util.communicator(), properties);
+    int bad_count = 0;
+    std::string error_message;
+    bool is_exodus_file = db != nullptr && db->ok(false, &error_message, &bad_count);
```

Comment:
Why does exodus appear here?  Can it be type agnostice?

Reply:
I'll remove references to exodus ... it's how it appeared in Framework IO 😬